### ISSUE TITLE
HttpClient proxyClient is not thread safe.

### ISSUE THREAD

**Issue by ppsimatikas** [mitre/HTTP-Proxy-Servlet#59]
I wanted to note here that I had a huge security issue occurring in my environment.
After tons and tons of investigation I have nailed it down to the ProxyServlet class under org.mitre.dsmiley.httpproxy.

A servlet instance is created only once when the server starts and it is used by all users.

Amazing explanation under BalusC's comment here http://stackoverflow.com/questions/3106452/how-do-servlets-work-instantiation-shared-variables-and-multithreading 

In your ProxyServlet  you have the HttpClient proxyClient created only once in the init() of the servlet.
This proxyClient instance is getting used by all users and it looks like it holds some variables from the one user when moving to the next one. I am using this proxy servlet to establish and grab login information from another web application that does NTLM authentication.  

A quick fix is that you generate a proxyClient instance for each user and each session...

So under ProxyServlet class within protected void service 

add these lines of code : 
 
            HttpClient theadSafeProxyClient = null;
            HttpSession session = servletRequest.getSession();

            if(session.getAttribute("proxyClient")!=null)
                theadSafeProxyClient = (HttpClient)session.getAttribute("proxyClient");
            else {
                BasicHttpParams hcParams = new BasicHttpParams();
                hcParams.setParameter("http.protocol.cookie-policy", "ignoreCookies");
                this.readConfigParam(hcParams, "http.protocol.handle-redirects", Boolean.class);
                theadSafeProxyClient = this.createHttpClient(hcParams);
                session.setAttribute("proxyClient", theadSafeProxyClient );
            }

before : 

proxyResponse1 = theadSafeproxyClient.execute(this.getTargetHost(servletRequest), (HttpRequest)proxyRequest);

**Comment by dsmiley** [comment]
Hi,
I am very aware that servlet instances are effectively singletons, and that this proxy servlet uses a HttpClient instance that is shared and might be used concurrently.  But HttpClient *should* be thread-safe; I'm certain of that, notwithstanding possible bugs.  _Maybe_ the proxy servlet is somehow picking a wrong/bad non-thread-safe instance under certain circumstances?  Or you found an HttpClient bug?  Which HttpClient version are you using?  If not the latest can you try the latest and see if the problem you found re-produces?

**Comment by ppsimatikas** [comment]
Apologies. My intention was not to question your skills and knowledge on servlets. I just wanted to have a fully explained issue report for other people that might face the same issue. I am using httpclient 4.4. I do not really think is a concurrency issue. It is more like multiple users are using the same HttpClient instance. That instance looks like it keeps some kind of session within. So one user is using the instance doing the proxy request and keeps some kind of session. Then the next user comes in and uses the same instance that provides session information of the previous user. That resulted to a huge security issue for my project.

**Comment by dsmiley** [comment]
Can you please point to something (some classes, etc.) where this occurs in HttpClient?  If this is happening, it's an HttpClient bug.  I'd like to investigate further.

**Comment by ppsimatikas** [comment]
I cannot identify exact class or method in HttpClient that this is happening. But it might not be a bug. It might actually be a feature of HttpClient to keep some kind of session for consecutive calls. I guess it is all about how the developer wish to use it.

**Comment by dsmiley** [comment]
Looking at HttpClient's docs, it declares itself to be thread-safe and that it should be shared.  And iit further specifies that state can optionally be used with HttpContext:   https://hc.apache.org/httpcomponents-client-4.4.x/tutorial/html/fundamentals.html#d5e213   (and see section 1.2.1 & 1.3).  The proxy servlet does not give an HttpContext, and so the particular overloaded execute() method the proxy uses will internally create one on the fly that only lives for the method call.  I saw this while stepping through it's code.

So at this point I'm not sure what to say.  Perhaps you can try version 4.5 of HttpClient, and also verify you are using a recent version of this proxy (i.e. the most recently released one in maven).

**Comment by systemsplanet** [comment]
The CookiePolicy.IGNORE_COOKIES uses 
   IgnoreSpec which extends CookieSpecBase which extends AbstractCookieSpec which uses a single shared for all requests:   private final Map<String, CookieAttributeHandler> attribHandlerMap;

I dont see any obvious reason why that would cause an issue since you should not be storing anything.
However,  the CookieSpec is not the only object used to process cookie headers before and after a request.

You may need  to set a CookieManager per request, or clear it between requests to avoid any leakage.

You may want to look into HttpCientBuilder and set cookieManagementDisabled(true). That disables the pre and post processing.

For CORS, you'll want to map the Origin header for OPTIONS in-flight requests.

In the past, I created a cookie spec that accepted all cookies, but it had issues, so I dont recommend.
mapping the host, port, url, Secure, HttpOnly is a lot of work.
Good luck!
PS.. you may want to all colon : as a cookie delimiter. I've seen some .Net clients use an older cookie spec
that says : is ok.

You've done a great job from what little I've see of your code.  
If it cant be done with one Class, then it's not worth doing!  :)

**Comment by dsmiley** [comment]
Thank for the clue as to what might be causing the problem! Unfortunately I'm too busy to work on fixes but I'd be happy to review Pull-Requests.

Curious; what does CORS have to do with this? I suspect that's feedback you are providing about proxy'ing in general and not specifically about the possible thread-safety bug?

**Comment by systemsplanet** [comment]
You are correct. Not specific to proxies, but probably a good enhancement to consider. CORS is used by modern web pages to avoid http connedction resets caused by too many requests to the same domain, and to display data from a third party site without violating the single-origin browser security model. It would not be too hard to add but you've done enough already.  Thanks again for sharing your time on this project.
